README.md
# MyKeyMouse
Blender keymapper addon - Map shortcuts on mouse button 4 and 5 for mouse with additional buttons.

**[Download latest](https://github.com/Pullusb/MyKeyMouse/archive/refs/heads/master.zip)** (right click, save Target as)  

For older blender 2.7 version go [here](https://github.com/Pullusb/SB_blender_addons_old_2_7)

If you want to support me, you can buy things on my [gumroad](https://pullusb.gumroad.com) or [blender market](https://blendermarket.com/creators/pullup) pages (other mean to support [here](http://www.samuelbernou.fr/donate)).

--------
  
### Description
  
Add '*view selection*' and '*view all*' shortcuts on mouse.
  
Keymapp a list of shortcut for almost all editor where:
- *mouse button 4* (often used as 'previous' action) is used to "get back", fit all elements in view. (like *home* button)
- *mouse button 5* (often used as 'next' action) is used to "go to", focus on selected items. (like *numpad period* button)

In the addons preferences you can choose to invert (button 4 to focus and button 5 to view all) :

It does NOT affect any original shortcut on keyboard.
<br/>
Usual location of button 4 and 5 on a mouse with basic extra buttons (this may vary a lot on different device)

![mouse with additional buttons 4 and 5](https://github.com/Pullusb/images_repo/blob/master/Mouse_button-4-5_zoom.png)
  
  
### Why ?
The location of *home* key and *numpad period* are far from both hands and force you to quit temporarily your rest positions, tending to break the workflow fluidity.
<br/>

### Other added keymaps

- *Ctrl + mouse button 5* Toggle isolate selection (*numpad slash* button)
<br/>

- *Shift + mouse button 4* Snap 3D cursor to selection  
- *Shift + mouse button 5* Snap selection to 3D cursor  
<br/>

- *Alt + mouse button 4* Jump to previous keyframe  
- *Alt + mouse button 5* Jump to next keyframe  
<br/>

- *Alt + Shift + mouse button 4* Jump to previous Marker  
- *Alt + Shift + mouse button 5* Jump to next Marker  
<br/>

- *Ctrl + Shift + alt + mouse mouse button 4* Origin to geometry
- *Ctrl + Shift + alt + mouse mouse button 5* Origin to cursor
<br/>

<!--
- *Ctrl + mouse button 4* Recenter the view on mouse cursor (like "Alt+F"), can be inverted in addon prefs  
- *Ctrl + mouse button 5* Toggle isolate selection (*numpad slash* button), can be inverted in addon prefs  
-->
 
Thanks to [Vincent Lamy for the base idea](https://www.nothing-is-3d.com/article22/blender-utiliser-les-boutons-lateraux-de-la-souris)


CHANGELOG.md
# changelog

1.3.0 - 2021-10-30:

- fix: bad mode in for snap selection to cursor / cursor to selected for Grease Pencil
- code: converted to multifile addons and added Changelog
- code: transfered 2.79 version to another repo

1.2.1 - 2020-12-20:

- fix: keymap unregister

1.2.0 - 2020-12-16:

- new shortcut: jump prev/next marker with `shift+alt+mouse4/5` (in every editor)

1.1.1 - 2020-11-28:

- feat: added cursor to selected and selection to cursor for grease pencil
- fix: Bug in keymap register that could affect other addon's keymap
- better UI for infos in addon prefs
- readme: changelog formating updated

1.0.0 - 13-09-2019:

- invert preference work only on the focus/view all shortcut
- fixed bug with keyframe jumping
- new shortcut : move origin point

0.0.9 - 02-06-2019:

- full change of keymap setting:
  - inverting default shortchut to a more logical setup (prev = going back, next = going to).
  - suppressing all preferences that overcomplicated the addon, letting only the possibility to invert everything (maybe delete even this one  later).

0.0.8 - 19-02-2019:

- 2.8 version

0.0.7 - 05-06-2018:

- Added Alt combo to jump keyframe

0.0.5 - 22-04-2018:

- Fix bad keymap unregister
- Added modifiers functions:
  - Shift combo : cursor to selection and selection to cursor
  - Ctrl combo : access alternative view tweak

0.0.3 - 14-04-2018:

- By default shortcuts are now completely consistent (view_selected and view all everywhere)
- added addon preferences to invert buttons and customize shortcut in 3D view.


__init__.py
bl_info = {
    "name": "MyKeyMouse",
    "description": "Add 'view selected' and 'view all' actions to mouse buttons 4 and 5",
    "author": "Samuel Bernou",
    "version": (1, 3, 0),
    "blender": (2, 80, 0),
    "location": "Mouse button 4 (usually 'previous') and 5 (usually 'next') on almost all editors",
    "warning": "",
    "doc_url": "https://github.com/Pullusb/MyKeyMouse",
    "category": "Object" }

import bpy
import os

#keymap focus view selected (numpad_period) to mouse button 4 on all related editors
#keymap view all (home key) to mouse button 5 on all related editors

#customisation
#3D view only :
#  Button 4 can view center pick on mouse instead (really a mouse related shortcut)
#  Button 5 can do local view instead (already have shit+C accessible for view all)


addon_keymaps = []
def register_keymaps():
    preferences = bpy.context.preferences
    addon_prefs = preferences.addons[__name__].preferences

    #mouse 4 and 5 often correspond to previous and next on mouse device
    key_prev = "BUTTON4MOUSE"#prev
    key_next = "BUTTON5MOUSE"#next

    if addon_prefs.mkmouse_invert_focus == True:
        key_focus = "BUTTON4MOUSE"
        key_home = "BUTTON5MOUSE"
    else:#default
        key_focus = "BUTTON5MOUSE"
        key_home = "BUTTON4MOUSE"

    shortcuts_items = [

    ##button 4 - focus selection
    ["3D View", "VIEW_3D", "view3d.view_selected", key_focus],
    ["Graph Editor", "GRAPH_EDITOR", "graph.view_selected", key_focus],
    ["Dopesheet", 'DOPESHEET_EDITOR', "action.view_selected", key_focus],
    ["Sequencer", 'SEQUENCE_EDITOR', "sequencer.view_selected", key_focus],
    ["Node Editor", "NODE_EDITOR", "node.view_selected", key_focus],
    ["NLA Editor", "NLA_EDITOR", "nla.view_selected", key_focus],
    ["Clip Editor", "CLIP_EDITOR", "clip.view_selected", key_focus],
    ["Image", "IMAGE_EDITOR", "image.view_selected", key_focus],
    ["Outliner", "OUTLINER", "outliner.show_active", key_focus],

    ##button 5 - view all
    ["3D View", "VIEW_3D", "view3d.view_all", key_home],
    ["Graph Editor", "GRAPH_EDITOR", "graph.view_all", key_home],
    ["Image", "IMAGE_EDITOR", "image.view_all", key_home],
    ["Node Editor", "NODE_EDITOR", "node.view_all", key_home],
    ["Dopesheet", "DOPESHEET_EDITOR", "action.view_all", key_home],
    ["NLA Editor", "NLA_EDITOR", "nla.view_all", key_home],
    ["Sequencer", "SEQUENCE_EDITOR", "sequencer.view_all", key_home],
    ["SequencerPreview", "SEQUENCE_EDITOR", "sequencer.view_all_preview", key_home],
    ["Clip Editor", "CLIP_EDITOR", "clip.view_all", key_home],
    ["Clip Graph Editor", "CLIP_EDITOR", "clip.graph_view_all", key_home],
    ["Clip Dopesheet Editor", "CLIP_EDITOR", "clip.dopesheet_view_all", key_home],

    #not exists anymore in 2.8
    # ["Timeline", "TIMELINE", "time.view_all", key_home],
    #["Logic Editor", "LOGIC_EDITOR", "logic.view_all", key_home],

    ]

    shortcuts_items.append(["3D View", "VIEW_3D", "view3d.localview", key_next, True, False, False])

    # Snapping utility with shift (cursor to selection and selection to selected)
    shortcuts_items.append(["3D View", "VIEW_3D", "view3d.snap_cursor_to_selected", key_prev, False, True, False])
    shortcuts_items.append(["3D View", "VIEW_3D", "view3d.snap_selected_to_cursor", key_next, False, True, False])
    
    # Same with GPencil (that have a different operator)
    shortcuts_items.append(["Grease Pencil Stroke Edit Mode", "EMPTY", "gpencil.snap_cursor_to_selected", key_prev, False, True, False])
    shortcuts_items.append(["Grease Pencil Stroke Edit Mode", "EMPTY", "gpencil.snap_to_cursor", key_next, False, True, False])


    ## 2.79 : 3D view > keymap view center pick on mouse (Alt+F) Changed to 'alt + MMB' to match early 2.8 settings
    #if LooseVersion(str(bpy.app.version)) < LooseVersion('(2, 80, 0)'):#only set for 2.79
    # shortcuts_items.append(["3D View", "VIEW_3D", "view3d.view_center_pick", "MIDDLEMOUSE", False, False, True])

    ## appending all keymap from above list
    addon = bpy.context.window_manager.keyconfigs.addon
    for item in shortcuts_items:
        #print(item)# printing items for debug
        km = addon.keymaps.new(name = item[0], space_type = item[1])
        if len(item) > 4:#contain modifiers keys
            kmi = km.keymap_items.new(item[2], type = item[3], value = "PRESS", ctrl=item[4], shift=item[5],alt=item[6])
        else:
            kmi = km.keymap_items.new(item[2], type = item[3], value = "PRESS")
        
        # print(item[2],'-->', kmi)
        addon_keymaps.append((km, kmi))

    # Moving origin point utility, cursor/geometry. (hold properties)
    km = addon.keymaps.new(name = "3D View", space_type = "VIEW_3D")
    kmi = km.keymap_items.new("object.origin_set", type = key_next, value = "PRESS", ctrl = True, shift = True, alt = True)
    kmi.properties.type = 'ORIGIN_CURSOR'
    addon_keymaps.append((km, kmi))
    
    kmi = km.keymap_items.new("object.origin_set", type = key_prev, value = "PRESS", ctrl = True, shift = True, alt = True)
    kmi.properties.type = 'ORIGIN_GEOMETRY'
    addon_keymaps.append((km, kmi))

    # Combo keyframe jump with alt special case (hold properties). Button BUTTON6MOUSE et 7 not working with logitech software...
    km = addon.keymaps.new(name = "Window", space_type = "EMPTY")
    kmi = km.keymap_items.new("screen.keyframe_jump", type = key_prev, value = "PRESS", alt = True)
    kmi.properties.next = False
    addon_keymaps.append((km, kmi))

    kmi = km.keymap_items.new("screen.keyframe_jump", type = key_next, value = "PRESS", alt = True)
    kmi.properties.next = True
    addon_keymaps.append((km, kmi))
    
    ## jump to marker
    kmi = km.keymap_items.new("screen.marker_jump", type = key_prev, value = "PRESS", alt = True, shift = True)
    kmi.properties.next = False
    addon_keymaps.append((km, kmi))

    kmi = km.keymap_items.new("screen.marker_jump", type = key_next, value = "PRESS", alt = True, shift = True)
    kmi.properties.next = True
    addon_keymaps.append((km, kmi))

###---user pref

class My_key_mouse_addon_pref(bpy.types.AddonPreferences):
    bl_idname = __name__

    mkmouse_invert_focus : bpy.props.BoolProperty(
        name="Invert (prev button = view selected, next button = view all. (Save prefs and restart Blender to apply changes)",
        default=False,
        )
    '''
    mkmouse_viewport_local_view : bpy.props.BoolProperty(
        name="Use local view (instead of view all)",
        default=False,
        )#"Use local view (like numpad slash) instead of view all (else combine ctrl)"
    mkmouse_viewport_center_on_mouse : bpy.props.BoolProperty(
        name="Use centering view on mouse (instead of view selected)",
        default=False,
        )#"Use centering view on mouse (like Alt+F) instead of view selected (else combine ctrl)"
    '''

    def draw(self, context):
        layout = self.layout
        box = layout.box()

        box.label(
            text="Focus control (all editor) :")
        box.label(
            text="mouse Prev button = view all")
        box.label(
            text="mouse Next button = view selected")
        # box.label(text="Customization ")
        box.prop(self, "mkmouse_invert_focus")
        

        box = layout.box()
        box.label(text="Only in 3D viewport:")
        box.label(text="Ctrl + mouse Next button = Use local view (like numpad slash)")
        #box.label(text="Ctrl + mouse Prev button = Unnassigned")
        # layout.separator()
        box.label(text="Cursor Snapping:")
        box.label(text="Shift + mouse Prev button = 3D cursor to selection")
        box.label(text="Shift + mouse Next button = selection to 3D cursor")
        # layout.separator()
        box.label(text="Origin change:")
        box.label(text="Ctrl + Shift + alt + mouse Prev button = Origin to geometry")
        box.label(text="Ctrl + Shift + alt + mouse Next button = Origin to cursor")

        # layout.label(text="Alt + middle mouse button = centering view on mouse (default shortcut in 2.8) as with Alt+F ")

        box = layout.box()
        box.label(text="Timeline related (all editor):")
        box.label(text="Alt + mouse Prev button = jump to prev keyframe")
        box.label(text="Alt + mouse Next button = jump to next keyframe")

        box.label(text="Alt + Shift + mouse Prev button = jump to prev marker")
        box.label(text="Alt + Shift + mouse Next button = jump to next marker")

        '''
        layout.label(
            text="Options to swap calls in 3D viewport (accessible with 'ctrl' modifier):")
        layout.prop(self, "mkmouse_viewport_center_on_mouse")
        layout.prop(self, "mkmouse_viewport_local_view")
        '''

def unregister_keymaps():
    # wm = bpy.context.window_manager
    for km, kmi in addon_keymaps:
        km.keymap_items.remove(kmi)
    addon_keymaps.clear()


## note : register via bpy.utils.register_module(__name__) Fails !
## addons prefs isn't loaded when keymap_register try to access it.

def register():
    if bpy.app.background:
        return
    bpy.utils.register_class(My_key_mouse_addon_pref)
    register_keymaps()

def unregister():
    if bpy.app.background:
        return
    unregister_keymaps()
    bpy.utils.unregister_class(My_key_mouse_addon_pref)

if __name__ == "__main__":
    register()


MyKeyMouse_279.py
bl_info = {
    "name": "MyKeyMouse",
    "description": "Add 'view all' and 'view selected' actions to mouse buttons 4 and 5",
    "author": "Samuel Bernou",
    "version": (0, 0, 9),
    "blender": (2, 79, 0),
    "location": "Mouse button 4 (usually 'previous') and 5 (usually 'next') on almost all editors",
    "warning": "",
    "wiki_url": "",
    "category": "Object" }

import bpy
import os
#from distutils.version import LooseVersion

'''
keymap view all (home key) to mouse button 4 on all related editors
keymap focus view selected (numpad_period) to mouse button 5 on all related editors
plus some special customisation

#available mouse button codenames
KeyMapItems.new(): keyword argument "type" : ('NONE', 'LEFTMOUSE', 'MIDDLEMOUSE', 'RIGHTMOUSE', 'BUTTON4MOUSE', 'BUTTON5MOUSE', 'BUTTON6MOUSE', 'BUTTON7MOUSE', 'ACTIONMOUSE', 'SELECTMOUSE', 'PEN', 'ERASER', 'MOUSEMOVE', 'INBETWEEN_MOUSEMOVE', 'TR)
'BUTTON6MOUSE', 'BUTTON7MOUSE' dont seem to work on windows
'''


addon_keymaps = []
def register_keymaps():
    user_preferences = bpy.context.user_preferences
    addon_prefs = user_preferences.addons[__name__].preferences

    #mouse 4 and 5 often correspond to previous and next on mouse device
    if addon_prefs.mkmouse_invert_buttons == True:
        key_one = "BUTTON4MOUSE"
        key_two = "BUTTON5MOUSE"
    else:#default
        key_one = "BUTTON5MOUSE"
        key_two = "BUTTON4MOUSE"

    shortcuts_items = [
 
    ##button 4 - focus selection
    ["3D View", "VIEW_3D", "view3d.view_selected", key_one],
    ["Graph Editor", "GRAPH_EDITOR", "graph.view_selected", key_one],
    ["Dopesheet", 'DOPESHEET_EDITOR', "action.view_selected", key_one],
    ["Sequencer", 'SEQUENCE_EDITOR', "sequencer.view_selected", key_one],
    ["Node Editor", "NODE_EDITOR", "node.view_selected", key_one],
    ["NLA Editor", "NLA_EDITOR", "nla.view_selected", key_one],
    ["Clip Editor", "CLIP_EDITOR", "clip.view_selected", key_one],
    ["Image", "IMAGE_EDITOR", "image.view_selected", key_one],
    ["Outliner", "OUTLINER", "outliner.show_active", key_one],

    ##button 5 - view all
    ["3D View", "VIEW_3D", "view3d.view_all", key_two],
    ["Timeline", "TIMELINE", "time.view_all", key_two],
    ["Graph Editor", "GRAPH_EDITOR", "graph.view_all", key_two],
    ["Image", "IMAGE_EDITOR", "image.view_all", key_two],
    ["Node Editor", "NODE_EDITOR", "node.view_all", key_two],
    ["Dopesheet", "DOPESHEET_EDITOR", "action.view_all", key_two],
    ["NLA Editor", "NLA_EDITOR", "nla.view_all", key_two],
    ["Sequencer", "SEQUENCE_EDITOR", "sequencer.view_all", key_two],
    ["SequencerPreview", "SEQUENCE_EDITOR", "sequencer.view_all_preview", key_two],
    ["Logic Editor", "LOGIC_EDITOR", "logic.view_all", key_two],
    ["Clip Editor", "CLIP_EDITOR", "clip.view_all", key_two],
    ["Clip Graph Editor", "CLIP_EDITOR", "clip.graph_view_all", key_two],
    ["Clip Dopesheet Editor", "CLIP_EDITOR", "clip.dopesheet_view_all", key_two],
    
    #["Frames", "EMPTY", "screen.keyframe_jump", key_one, False, False, True]
    
    ]

    shortcuts_items.append(["3D View", "VIEW_3D", "view3d.localview", key_one, True, False, False])

    # Snapping utility with shift (cursor to selection and selection to selected)
    shortcuts_items.append(["3D View", "VIEW_3D", "view3d.snap_cursor_to_selected", key_two, False, True, False])
    shortcuts_items.append(["3D View", "VIEW_3D", "view3d.snap_selected_to_cursor", key_one, False, True, False])

    
    ##3D view > keymap view center pick on mouse (Alt+F) Changed to 'alt + MMB' to match early 2.8 settings
    #if LooseVersion(str(bpy.app.version)) < LooseVersion('(2, 80, 0)'):#only set for 2.79
    shortcuts_items.append(["3D View", "VIEW_3D", "view3d.view_center_pick", "MIDDLEMOUSE", False, False, True])

    ## appending all keymap from above list
    addon = bpy.context.window_manager.keyconfigs.addon
    for item in shortcuts_items:
        #print(item)# printing items for debug
        km = addon.keymaps.new(name = item[0], space_type = item[1])
        if len(item) > 4:#contain modifiers keys
            kmi = km.keymap_items.new(item[2], type = item[3], value = "PRESS", ctrl=item[4], shift=item[5],alt=item[6])
        else:
            kmi = km.keymap_items.new(item[2], type = item[3], value = "PRESS")
        addon_keymaps.append(km)


        #special case (because hold properties), combo keyframe jump with alt (button 6 et 7 not working with logitech software...)
        km = addon.keymaps.new(name = "Frames", space_type = "EMPTY")
        #kmi = km.keymap_items.new("screen.keyframe_jump", type = "BUTTON6MOUSE", value = "PRESS")#mouse button above 5 aren't recognize on logitech mouse on windaube
        kmi = km.keymap_items.new("screen.keyframe_jump", type = key_one, value = "PRESS", alt = True)
        kmi.properties.next = False
        #kmi = km.keymap_items.new("screen.keyframe_jump", type = "BUTTON7MOUSE", value = "PRESS")#mouse button above 5 aren't recognize on logitech mouse on windaube
        kmi = km.keymap_items.new("screen.keyframe_jump", type = key_two, value = "PRESS", alt = True)
        kmi.properties.next = True




###---user pref 

class My_key_mouse_addon_pref(bpy.types.AddonPreferences):
    bl_idname = __name__

    mkmouse_invert_buttons = bpy.props.BoolProperty(
        name="Invert buttons everywhere (prev button = view all, next button = view selected)",
        default=False,
        )
    mkmouse_viewport_local_view = bpy.props.BoolProperty(
        name="Use local view (instead of view all)",
        default=False,
        )#"Use local view (like numpad slash) instead of view all (else combine ctrl)"
    '''
    mkmouse_viewport_center_on_mouse = bpy.props.BoolProperty(
        name="Use centering view on mouse (instead of view selected)",
        default=False,
        )#"Use centering view on mouse (like Alt+F) instead of view selected (else combine ctrl)"
    '''

    def draw(self, context):
        layout = self.layout
        layout.label(
            text="In all editor :")
        layout.label(
            text="mouse Prev button = view selected")
        layout.label(
            text="mouse Next button = view all")
        layout.label(text="")

        layout.label(
            text="Only in 3D viewport:")
        
        layout.label(text="Ctrl + mouse Next button = Use local view (like numpad slash)")
        #layout.label(text="Ctrl + mouse Prev button = Use centering view on mouse ")

        layout.label("Cursor Snapping:")
        layout.label("Shift + mouse Prev button = 3D cursor to selection")
        layout.label("Shift + mouse Next button = selection to 3D cursor")
        layout.label(text="")

        layout.label("Extra:")
        layout.label("Alt + middle mouse button = centering view on mouse (default shortcut in 2.8) as with Alt+F ")
        layout.label("Alt + mouse Prev button = jump to prev keyframe ")
        layout.label("Alt + mouse Next button = jump to next keyframe ")
        layout.label(text="")
        layout.label(
            text="Customization (save settings and restart Blender to apply changes)")
        layout.prop(self, "mkmouse_invert_buttons")
        '''
        layout.label(
            text="Options to swap calls in 3D viewport (accessible with 'ctrl' modifier):") 
        layout.prop(self, "mkmouse_viewport_center_on_mouse")
        layout.prop(self, "mkmouse_viewport_local_view")
        '''

def unregister_keymaps():
    wm = bpy.context.window_manager
    for km in addon_keymaps:
        for kmi in km.keymap_items:
            km.keymap_items.remove(kmi)
        ## Can't (and supposedly shouldn't ) suppress original category name...
        # wm.keyconfigs.addon.keymaps.remove(km)
    addon_keymaps.clear()



## note : register via bpy.utils.register_module(__name__) Fails !
## addons prefs isn't loaded when keymap_register try to access it.

def register():
    if not bpy.app.background:
        bpy.utils.register_class(My_key_mouse_addon_pref)
        register_keymaps()

def unregister():
    if not bpy.app.background:
        unregister_keymaps()
        bpy.utils.unregister_class(My_key_mouse_addon_pref)

if __name__ == "__main__":
    register()

MyKeyMouse.py
bl_info = {
    "name": "MyKeyMouse",
    "description": "Add 'view selected' and 'view all' actions to mouse buttons 4 and 5",
    "author": "Samuel Bernou",
    "version": (1, 2, 2),
    "blender": (2, 80, 0),
    "location": "Mouse button 4 (usually 'previous') and 5 (usually 'next') on almost all editors",
    "warning": "",
    "doc_url": "https://github.com/Pullusb/MyKeyMouse",
    "category": "Object" }

import bpy
import os

#keymap focus view selected (numpad_period) to mouse button 4 on all related editors
#keymap view all (home key) to mouse button 5 on all related editors

#customisation
#3D view only :
#  Button 4 can view center pick on mouse instead (really a mouse related shortcut)
#  Button 5 can do local view instead (already have shit+C accessible for view all)


addon_keymaps = []
def register_keymaps():
    preferences = bpy.context.preferences
    addon_prefs = preferences.addons[__name__].preferences

    #mouse 4 and 5 often correspond to previous and next on mouse device
    key_prev = "BUTTON4MOUSE"#prev
    key_next = "BUTTON5MOUSE"#next

    if addon_prefs.mkmouse_invert_focus == True:
        key_focus = "BUTTON4MOUSE"
        key_home = "BUTTON5MOUSE"
    else:#default
        key_focus = "BUTTON5MOUSE"
        key_home = "BUTTON4MOUSE"

    shortcuts_items = [

    ##button 4 - focus selection
    ["3D View", "VIEW_3D", "view3d.view_selected", key_focus],
    ["Graph Editor", "GRAPH_EDITOR", "graph.view_selected", key_focus],
    ["Dopesheet", 'DOPESHEET_EDITOR', "action.view_selected", key_focus],
    ["Sequencer", 'SEQUENCE_EDITOR', "sequencer.view_selected", key_focus],
    ["Node Editor", "NODE_EDITOR", "node.view_selected", key_focus],
    ["NLA Editor", "NLA_EDITOR", "nla.view_selected", key_focus],
    ["Clip Editor", "CLIP_EDITOR", "clip.view_selected", key_focus],
    ["Image", "IMAGE_EDITOR", "image.view_selected", key_focus],
    ["Outliner", "OUTLINER", "outliner.show_active", key_focus],

    ##button 5 - view all
    ["3D View", "VIEW_3D", "view3d.view_all", key_home],
    ["Graph Editor", "GRAPH_EDITOR", "graph.view_all", key_home],
    ["Image", "IMAGE_EDITOR", "image.view_all", key_home],
    ["Node Editor", "NODE_EDITOR", "node.view_all", key_home],
    ["Dopesheet", "DOPESHEET_EDITOR", "action.view_all", key_home],
    ["NLA Editor", "NLA_EDITOR", "nla.view_all", key_home],
    ["Sequencer", "SEQUENCE_EDITOR", "sequencer.view_all", key_home],
    ["SequencerPreview", "SEQUENCE_EDITOR", "sequencer.view_all_preview", key_home],
    ["Clip Editor", "CLIP_EDITOR", "clip.view_all", key_home],
    ["Clip Graph Editor", "CLIP_EDITOR", "clip.graph_view_all", key_home],
    ["Clip Dopesheet Editor", "CLIP_EDITOR", "clip.dopesheet_view_all", key_home],

    #not exists anymore in 2.8
    # ["Timeline", "TIMELINE", "time.view_all", key_home],
    #["Logic Editor", "LOGIC_EDITOR", "logic.view_all", key_home],

    ]

    shortcuts_items.append(["3D View", "VIEW_3D", "view3d.localview", key_next, True, False, False])

    # Snapping utility with shift (cursor to selection and selection to selected)
    shortcuts_items.append(["3D View", "VIEW_3D", "view3d.snap_cursor_to_selected", key_prev, False, True, False])
    shortcuts_items.append(["3D View", "VIEW_3D", "view3d.snap_selected_to_cursor", key_next, False, True, False])
    
    # Same with GPencil (that have a different operator)
    shortcuts_items.append(["Grease Pencil", "EMPTY", "gpencil.snap_cursor_to_selected", key_prev, False, True, False])
    shortcuts_items.append(["Grease Pencil", "EMPTY", "gpencil.snap_selected_to_cursor", key_next, False, True, False])


    ## 2.79 : 3D view > keymap view center pick on mouse (Alt+F) Changed to 'alt + MMB' to match early 2.8 settings
    #if LooseVersion(str(bpy.app.version)) < LooseVersion('(2, 80, 0)'):#only set for 2.79
    # shortcuts_items.append(["3D View", "VIEW_3D", "view3d.view_center_pick", "MIDDLEMOUSE", False, False, True])

    ## appending all keymap from above list
    addon = bpy.context.window_manager.keyconfigs.addon
    for item in shortcuts_items:
        #print(item)# printing items for debug
        km = addon.keymaps.new(name = item[0], space_type = item[1])
        if len(item) > 4:#contain modifiers keys
            kmi = km.keymap_items.new(item[2], type = item[3], value = "PRESS", ctrl=item[4], shift=item[5],alt=item[6])
        else:
            kmi = km.keymap_items.new(item[2], type = item[3], value = "PRESS")
        
        # print(item[2],'-->', kmi)
        addon_keymaps.append((km, kmi))

    # Moving origin point utility, cursor/geometry. (hold properties)
    km = addon.keymaps.new(name = "3D View", space_type = "VIEW_3D")
    kmi = km.keymap_items.new("object.origin_set", type = key_next, value = "PRESS", ctrl = True, shift = True, alt = True)
    kmi.properties.type = 'ORIGIN_CURSOR'
    addon_keymaps.append((km, kmi))
    
    kmi = km.keymap_items.new("object.origin_set", type = key_prev, value = "PRESS", ctrl = True, shift = True, alt = True)
    kmi.properties.type = 'ORIGIN_GEOMETRY'
    addon_keymaps.append((km, kmi))

    # Combo keyframe jump with alt special case (hold properties). Button BUTTON6MOUSE et 7 not working with logitech software...
    km = addon.keymaps.new(name = "Window", space_type = "EMPTY")
    kmi = km.keymap_items.new("screen.keyframe_jump", type = key_prev, value = "PRESS", alt = True)
    kmi.properties.next = False
    addon_keymaps.append((km, kmi))

    kmi = km.keymap_items.new("screen.keyframe_jump", type = key_next, value = "PRESS", alt = True)
    kmi.properties.next = True
    addon_keymaps.append((km, kmi))
    
    ## jump to marker
    kmi = km.keymap_items.new("screen.marker_jump", type = key_prev, value = "PRESS", alt = True, shift = True)
    kmi.properties.next = False
    addon_keymaps.append((km, kmi))

    kmi = km.keymap_items.new("screen.marker_jump", type = key_next, value = "PRESS", alt = True, shift = True)
    kmi.properties.next = True
    addon_keymaps.append((km, kmi))

###---user pref

class My_key_mouse_addon_pref(bpy.types.AddonPreferences):
    bl_idname = __name__

    mkmouse_invert_focus : bpy.props.BoolProperty(
        name="Invert (prev button = view selected, next button = view all. (Save prefs and restart Blender to apply changes)",
        default=False,
        )
    '''
    mkmouse_viewport_local_view : bpy.props.BoolProperty(
        name="Use local view (instead of view all)",
        default=False,
        )#"Use local view (like numpad slash) instead of view all (else combine ctrl)"
    mkmouse_viewport_center_on_mouse : bpy.props.BoolProperty(
        name="Use centering view on mouse (instead of view selected)",
        default=False,
        )#"Use centering view on mouse (like Alt+F) instead of view selected (else combine ctrl)"
    '''

    def draw(self, context):
        layout = self.layout
        box = layout.box()

        box.label(
            text="Focus control (all editor) :")
        box.label(
            text="mouse Prev button = view all")
        box.label(
            text="mouse Next button = view selected")
        # box.label(text="Customization ")
        box.prop(self, "mkmouse_invert_focus")
        

        box = layout.box()
        box.label(text="Only in 3D viewport:")
        box.label(text="Ctrl + mouse Next button = Use local view (like numpad slash)")
        #box.label(text="Ctrl + mouse Prev button = Unnassigned")
        # layout.separator()
        box.label(text="Cursor Snapping:")
        box.label(text="Shift + mouse Prev button = 3D cursor to selection")
        box.label(text="Shift + mouse Next button = selection to 3D cursor")
        # layout.separator()
        box.label(text="Origin change:")
        box.label(text="Ctrl + Shift + alt + mouse Prev button = Origin to geometry")
        box.label(text="Ctrl + Shift + alt + mouse Next button = Origin to cursor")

        # layout.label(text="Alt + middle mouse button = centering view on mouse (default shortcut in 2.8) as with Alt+F ")

        box = layout.box()
        box.label(text="Timeline related (all editor):")
        box.label(text="Alt + mouse Prev button = jump to prev keyframe")
        box.label(text="Alt + mouse Next button = jump to next keyframe")

        box.label(text="Alt + Shift + mouse Prev button = jump to prev marker")
        box.label(text="Alt + Shift + mouse Next button = jump to next marker")

        '''
        layout.label(
            text="Options to swap calls in 3D viewport (accessible with 'ctrl' modifier):")
        layout.prop(self, "mkmouse_viewport_center_on_mouse")
        layout.prop(self, "mkmouse_viewport_local_view")
        '''

def unregister_keymaps():
    # wm = bpy.context.window_manager
    for km, kmi in addon_keymaps:
        km.keymap_items.remove(kmi)
    addon_keymaps.clear()


## note : register via bpy.utils.register_module(__name__) Fails !
## addons prefs isn't loaded when keymap_register try to access it.

def register():
    if bpy.app.background:
        return
    bpy.utils.register_class(My_key_mouse_addon_pref)
    register_keymaps()

def unregister():
    if bpy.app.background:
        return
    unregister_keymaps()
    bpy.utils.unregister_class(My_key_mouse_addon_pref)

if __name__ == "__main__":
    register()
